usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -2479,6 +2521,14 @@ CK_RV ep11tok_init(STDLL_TokData_t * tokdata, CK_SLOT_ID SlotNumber,
     if (ep11_data == NULL)
         return CKR_HOST_MEMORY;
 
+    if (pthread_rwlock_init(&ep11_data->target_rwlock, NULL) != 0) {
```

Comment:
well, we don't care about this flag at many places! That would be an effort on its own. We certainly could spend that effort at some time, but I would prefer to not change this here in this PR.

Reply:
Agreed.  Found some other places where this is not considered.  Guess that is another cleanup at some point.

Reply:
However, we should probably check for CKF_LIBRARY_CANT_CREATE_OS_THREADS and return CKR_NEED_TO_CREATE_THREADS if it is set. We do need to use OS threads for the event thread, so we can't work with CKF_LIBRARY_CANT_CREATE_OS_THREADS  being set.

Reply:
I have added a separate commit for this.

Reply:
Re CKF_OS_LOCKING_OK: 
The strange thing about this is that CK_C_INITIALIZE_ARGS provides user supplied functions for Mutex-type locks only. No RW locks. Certainly, a RW lock could be replaced by a mutex-type lock, but that would probably cause performance problems, because RW locks allow multiple concurrent readers. Also recursive use of  locks might be a problem. I guess this needs proper analysis and design in order to support application provided locking functions. Maybe RW-locks and recursive locks can be somehow implemented/simulated by use of mutexes...

Reply:
Well, I am not sure how PKCS#11 is supposed to work in that case anyway.  There are some features that seem inherently concurrent (event support being one) that you simply cannot implement without threads.  The reason for this flag might be managed engines like Java or .NET where you want to use managed threads instead of OS threads, but there is no way for a PKCS#11 implementation to get such a thread.  Sure, we could simulate read/write locking and recursive mutexes via wrappers using the provided callbacks (with quite a bit of performance penality - but actually we are missing a non-spinning notify-wait mechanism for this which is not provided by PKCS#11), but we cannot do anything about threading.

So I guess in the end we have two options:
1. Fail initialization if we are not allowed to create threads or if OS locking is not okay (i.e., if `flags != CKF_OS_LOCKING_OK`)
2. Restrict the library functionality to accommodate for user-provided restrictions.

Personally, I prefer 1 since that makes the functionality equal in all cases.

Reply:
If you search the internet for the locking and thread flags, it seems that this is not really relevant today anymore, especially not on Linux. It was meant for environments that use its own threading scheme, and thus needs own locking primitives. So I guess that for our case we could actually fail if the flags are set in a way that we cant work with.